src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,208 @@
+    return ec->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(msgBuf), sendFlags);
+}
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                       System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "OnMessageReceived: Failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+
+        // This should notify the tranfer object to abort transfer so it can send a status report across the exchange
+        // when we call ProcessOutputEvents below.
+        mTransfer.AbortTransfer(AsyncResponder::GetBdxStatusCodeFromChipError(err));
+    }
+    else if (!payloadHeader.HasMessageType(MessageType::BlockAckEOF))
+    {
+
+        // Almost every BDX message expect BlockAckEOF will follow up with a response on the exchange.
+        ec->WillSendMessage();
+    }
+
+    ProcessOutputEvents();
+    return err;
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnResponseTimeout, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    DestroySelf();
+}
+
+CHIP_ERROR AsyncResponder::Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                System::Clock::Timeout timeout)
+{
+    AsyncTransferFacilitator::Init(layer, exchangeCtx, timeout);
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
+    return CHIP_NO_ERROR;
+}
+
+void AsyncResponder::NotifyEventHandled(const TransferSession::OutputEvent & event, CHIP_ERROR status)
+{
+    ChipLogDetail(BDX, "NotifyEventHandled : Event %s Error %" CHIP_ERROR_FORMAT,
+                  TransferSession::OutputEvent::TypeToString(event.EventType), status.Format());
+
+    // If it's a message indicating either the end of the transfer or a timeout reported by the transfer session
+    // or an error occured, we need to call DestroySelf().
+    if (event.EventType == TransferSession::OutputEventType::kAckEOFReceived ||
+        event.EventType == TransferSession::OutputEventType::kInternalError ||
+        event.EventType == TransferSession::OutputEventType::kTransferTimeout ||
+        event.EventType == TransferSession::OutputEventType::kStatusReceived)
+    {
+
+        // Set the mDestroySelf flag to true so that when ProcessOutputEvents finishes processing all pending events, it
+        // will call DestroySelf() to clean up.
```

Comment:
This comment would be wholly unnecessary if the flag were named clearly; see naming suggestion.

Reply:
removed comment.